Upgrade stripe-java to 20.33.0

skress opened the pull request:
Upgrade plugin to Stripe API version https://stripe.com/docs/upgrades#2020-08-27.

All tests are successfully running, except of `testSuccessfulBankAccountPurchase` which was already broken.

I am not sure how you are defining the version number of this plugin. So what's currently missing is an update of README.md to reflect the updated API version. Should this be a new row and plugin version 7.2.y or will it remain at 7.1.y?

pierre commented:
> All tests are successfully running, except of `testSuccessfulBankAccountPurchase` which was already broken.

What is the error? CI tests are all passing against our sandbox.
 
> I am not sure how you are defining the version number of this plugin. So what's currently missing is an update of README.md to reflect the updated API version. Should this be a new row and plugin version 7.2.y or will it remain at 7.1.y?

Yes, let's bump it to 7.2.y, so we can provide further updates on the 7.1.y release train for companies who cannot upgrade right away.

Thanks!

skress commented:
Note that the following description is based on running `testSuccessfulBankAccountPurchase` against the current master branch, i.e. the current version of the Stripe plugin without my changes.

The error is

```
com.stripe.exception.InvalidRequestException: You may only specify one of these parameters: ["source", "payment_method"].; request-id: req_Oj4Bp9TeRtCYpT

	at com.stripe.net.LiveStripeResponseGetter.handleApiError(LiveStripeResponseGetter.java:157)
	at com.stripe.net.LiveStripeResponseGetter.request(LiveStripeResponseGetter.java:61)
	at com.stripe.net.ApiResource.request(ApiResource.java:179)
	at com.stripe.model.Customer.create(Customer.java:220)
	at org.killbill.billing.plugin.stripe.TestStripePaymentPluginApi.createStripeCustomerWithBankAccountAndCreditCard(TestStripePaymentPluginApi.java:790)
```

After commenting out https://github.com/killbill/killbill-stripe-plugin/blob/master/src/test/java/org/killbill/billing/plugin/stripe/TestStripePaymentPluginApi.java#L789 I still ran into another problem. But I just found out that this problem resulted from the Stripe public key being hardcoded here: https://github.com/killbill/killbill-stripe-plugin/blob/master/src/test/java/org/killbill/billing/plugin/stripe/TestStripePaymentPluginApi.java#L830

After changing the key to my public key I was finally blocked by 'ACH payments from DE accounts are not currently supported'. So I will never be able to run that test ...

Still the question remains why you do not run into the first problem.

skress commented:
I've updated the PR to change the version from 7.1.y to 7.2.y. I hope that I found all places where I needed to change the version.

Somehow your latest commit 'Add "ach_debit" payment method option only for USD' was also included in that change as well ... don't really know why and if/how I could revert that.

pierre commented:
> Note that the following description is based on running `testSuccessfulBankAccountPurchase` against the current master branch, i.e. the current version of the Stripe plugin without my changes.
> 
> The error is
> 
> ```
> com.stripe.exception.InvalidRequestException: You may only specify one of these parameters: ["source", "payment_method"].; request-id: req_Oj4Bp9TeRtCYpT
> 
> 	at com.stripe.net.LiveStripeResponseGetter.handleApiError(LiveStripeResponseGetter.java:157)
> 	at com.stripe.net.LiveStripeResponseGetter.request(LiveStripeResponseGetter.java:61)
> 	at com.stripe.net.ApiResource.request(ApiResource.java:179)
> 	at com.stripe.model.Customer.create(Customer.java:220)
> 	at org.killbill.billing.plugin.stripe.TestStripePaymentPluginApi.createStripeCustomerWithBankAccountAndCreditCard(TestStripePaymentPluginApi.java:790)
> ```

I just verified again that the tests all pass using `master`. Maybe my sandbox is configured a bit differently.

Could you reach out to Stripe support to see if there is something they need to enable on your account?

> After commenting out https://github.com/killbill/killbill-stripe-plugin/blob/master/src/test/java/org/killbill/billing/plugin/stripe/TestStripePaymentPluginApi.java#L789 I still ran into another problem. But I just found out that this problem resulted from the Stripe public key being hardcoded here: https://github.com/killbill/killbill-stripe-plugin/blob/master/src/test/java/org/killbill/billing/plugin/stripe/TestStripePaymentPluginApi.java#L830

Oops, good catch! Could you replace the hardcoded value with the following?

```
stripeConfigPropertiesConfigurationHandler.getConfigurable(super.context.getTenantId()).getPublicKey();
```

> After changing the key to my public key I was finally blocked by 'ACH payments from DE accounts are not currently supported'. So I will never be able to run that test ...

Got it. I took your branch and I'm having a few issues.

First one is in the test:

```
[ERROR] org.killbill.billing.plugin.stripe.TestStripePaymentPluginApi.testSuccessfulBankAccountPurchase  Time elapsed: 2.078 s  <<< FAILURE!
java.lang.NullPointerException
	at org.killbill.billing.plugin.stripe.TestStripePaymentPluginApi.createStripeCustomerWithBankAccountAndCreditCard(TestStripePaymentPluginApi.java:800)
```

If I add a NPE check, I now get a NPE in the code:

```
[ERROR] org.killbill.billing.plugin.stripe.TestStripePaymentPluginApi.testSuccessfulBankAccountPurchase  Time elapsed: 3.098 s  <<< FAILURE!
java.lang.NullPointerException
	at org.killbill.billing.plugin.stripe.StripePaymentPluginApi$4.execute(StripePaymentPluginApi.java:752)
	at org.killbill.billing.plugin.stripe.StripePaymentPluginApi$4.execute(StripePaymentPluginApi.java:725)
	at org.killbill.billing.plugin.stripe.StripePaymentPluginApi.executeInitialTransaction(StripePaymentPluginApi.java:810)
	at org.killbill.billing.plugin.stripe.StripePaymentPluginApi.executeInitialTransaction(StripePaymentPluginApi.java:724)
	at org.killbill.billing.plugin.stripe.StripePaymentPluginApi.purchasePayment(StripePaymentPluginApi.java:563)
```

The behavior of the new client seems to be quite different. Could you create a new dummy US Stripe sandbox account, so you can ensure all tests are passing?

> I've updated the PR to change the version from 7.1.y to 7.2.y. I hope that I found all places where I needed to change the version.

👍 

> Somehow your latest commit 'Add "ach_debit" payment method option only for USD' was also included in that change as well ... don't really know why and if/how I could revert that.

Could you bring your branch up-to-date with master?

skress commented:
The response that I got from Stripe does not really help. They say that the error that I am getting is correct, i.e. you can either specify "source" or "payment_method" but not both. There does not seem to be a configuration option to allow placing both options together.

Would it be possible that you run that one test again, then go to https://dashboard.stripe.com/test/logs and fetch the request ID of  the POST to /v1/customers and then ask the Stripe support why that request went through (and perhaps adding that other Stripe accounts get errors when placing "source" and "payment_method" in the same request)?

Just for reference, this is how it looks like when I run the test from the master branch (without my changes):

POST /v1/customers 400 ERR
API-Version 2019-12-03
Source Stripe/v1 JavaBindings/17.4.0

Request-POST-Text
{
  "source": "btok_1I6dWf2em0gu2nL95xPEiYzP",
  "payment_method": "pm_card_visa"
}
Response-Text
{
  "error": {
    "message": "You may only specify one of these parameters: [\"source\", \"payment_method\"].",
    "param": "source",
    "type": "invalid_request_error"
  }
}

pierre commented:
> The response that I got from Stripe does not really help. They say that the error that I am getting is correct, i.e. you can either specify "source" or "payment_method" but not both. There does not seem to be a configuration option to allow placing both options together.

😕 

> Would it be possible that you run that one test again, then go to https://dashboard.stripe.com/test/logs and fetch the request ID of the POST to /v1/customers and then ask the Stripe support why that request went through (and perhaps adding that other Stripe accounts get errors when placing "source" and "payment_method" in the same request)?

Unfortunately, we don't have a commercial relationship with Stripe, so we don't have access to support.

Here's what I am seeing:

## master

Response

```
POST        /v1/customers
Status      200 OK
ID          req_pR1ieJMn4a6jzg
Time        1/7/21, 12:26:17 PM
API version 2019-12-03
Source      Stripe/v1 JavaBindings/17.4.0
```

Request POST body

```
{
  "source": "btok_1I6xZE2em0gu2nL9nQbuvk0Q",
  "payment_method": "pm_card_visa"
}
```

## Your branch

Response

```
POST        /v1/customers
Status      200 OK
ID          req_DYudQyTOVtnBRf
Time        1/7/21, 12:20:21 PM
API version 2020-08-27
Source      Stripe/v1 JavaBindings/20.33.0
```

Request POST body

```
{
  "expand": {
    "0": "sources"
  },
  "source": "btok_1I6xTU2em0gu2nL99m0DybN2",
  "payment_method": "pm_card_visa"
}
```

Feel free to send this to support, telling them they can contact me directly if they need more information.

Now, taking a step back, this error only means that somehow you cannot add multiple payment methods in one go on your account when creating a Stripe Customer, so I don't think this is a blocker to merge this PR.

Can you just confirm that:

* Commenting out https://github.com/killbill/killbill-stripe-plugin/blob/fe750badb121ddd8f179d876b7d9ccfae45eda71/src/test/java/org/killbill/billing/plugin/stripe/TestStripePaymentPluginApi.java#L789 makes the suite pass on your US account, and that all other tests pass on your DE account (i.e. all except the bank account one)?
* You were able to run the tests marked as `Manual test` (e.g. `testSuccessful3DSAuthCapture`)?

If so, we should be good to go. Thank you for your patience, the latest iteration of the Stripe APIs has been quite challenging to integrate with, especially as we needed to provide backward compatibility here in the US 😓

skress commented:
> Now, taking a step back, this error only means that somehow you cannot add multiple payment methods in one go on your account when creating a Stripe Customer, so I don't think this is a blocker to merge this PR.
> 
> Can you just confirm that:
> 
> * Commenting out https://github.com/killbill/killbill-stripe-plugin/blob/fe750badb121ddd8f179d876b7d9ccfae45eda71/src/test/java/org/killbill/billing/plugin/stripe/TestStripePaymentPluginApi.java#L789
>    makes the suite pass on your US account, and that all other tests pass on your DE account (i.e. all except the bank account one)?
> * You were able to run the tests marked as `Manual test` (e.g. `testSuccessful3DSAuthCapture`)?
> 
> If so, we should be good to go. 

Yes, I have run all tests including those marked as manual on my US and DE accounts. When commenting out that one line, all tests including the manual ones passed on the US account. Additionally all tests except this one passed on the DE account as well (this one failed as ACH payments are not possible on DE accounts).

Nevertheless I have committed a small change to the test (and successfully re-run all tests again). As I understood the intention of that test, you wanted to have a second payment method on the account to verify that you support multiple payment method types per account. I have removed adding the credit card from the call when creating the customer and added separate calls to add the credit card immediately after creating the customer. I have checked it in Stripe's dashboard and the customer had two payment methods attached. So it is working for me. Hopefully it works in your environment as it does in mine.

If the intent of the test is still captured, I guess this might save us some future trouble if anyone else is running those tests with newer Stripe accounts ...

> Thank you for your patience, the latest iteration of the Stripe APIs has been quite challenging to integrate with, especially as we needed to provide backward compatibility here in the US 😓

No worries, I am happy that KB exists and that you implemented the Stripe plugin in the first place ;-)

pierre reviewed (approved):
Thank you!

pierre commented:
7.2.0 has been released.
